File: src/batchNorm2d.py
import torch
import numpy as np
import os

# **Configuration Parameters**
OUT_C = 3     
H_OUT = 10    
W_OUT = 12    

# **Ensure the `data` directory exists**
os.makedirs("data", exist_ok=True)

# **1. Generate Random Data**
torch.manual_seed(42)
inBuffer = torch.rand(1, OUT_C, H_OUT, W_OUT, dtype=torch.float32)  # Random input
gamma = torch.rand(OUT_C, dtype=torch.float32)   # Scale parameter (γ)
beta = torch.rand(OUT_C, dtype=torch.float32)    # Bias parameter (β)
running_mean = torch.rand(OUT_C, dtype=torch.float32)  # Running mean (used in inference)
running_var = torch.rand(OUT_C, dtype=torch.float32)   # Running variance (used in inference)

# **2. Create PyTorch BatchNorm2d Layer**
batchnorm = torch.nn.BatchNorm2d(OUT_C, affine=True, track_running_stats=True)
batchnorm.weight.data = gamma
batchnorm.bias.data = beta
batchnorm.running_mean = running_mean
batchnorm.running_var = running_var

# **3. Compute BatchNorm**
batchnorm.eval()  # Set to inference mode
out_torch = batchnorm(inBuffer).squeeze(0).detach().numpy()  # Remove batch dimension

# **4. Save Data (in `.npy` format)**
np.save("data/inBuffer_bn2d.npy", inBuffer.numpy())   # Input
np.save("data/gamma_bn2d.npy", gamma.numpy())         # γ (Scale)
np.save("data/beta_bn2d.npy", beta.numpy())           # β (Bias)
np.save("data/running_mean_bn2d.npy", running_mean.numpy())  # Running mean
np.save("data/running_var_bn2d.npy", running_var.numpy())   # Running variance
np.save("data/out_torch_bn2d.npy", out_torch)         # PyTorch computation result

# **5. Save Data (in `.txt` format)**
np.savetxt("data/inBuffer_bn2d.txt", inBuffer.numpy().flatten(), fmt="%.6f")
np.savetxt("data/gamma_bn2d.txt", gamma.numpy(), fmt="%.6f")
np.savetxt("data/beta_bn2d.txt", beta.numpy(), fmt="%.6f")
np.savetxt("data/running_mean_bn2d.txt", running_mean.numpy(), fmt="%.6f")
np.savetxt("data/running_var_bn2d.txt", running_var.numpy(), fmt="%.6f")
np.savetxt("data/out_torch_bn2d.txt", out_torch.flatten(), fmt="%.6f")

print("BatchNorm2d data has been saved to the `data/` directory, including `.npy` and `.txt` formats.")

File: src/batchNorm2d.cpp
/* 
This C++ implementation of Batch Normalization explicitly calculate the mean, variance, and standard deviation for normalization using the formula from
https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html
This involves square root and division operations, which are computationally expensive and inefficient for hardware implementation.
In contrast, many hardware-friendly implementations assume that the input is already normalized, meaning the mean and variance are preprocessed, 
and the network directly scales the input with learned gamma (scale) and beta (bias) parameters. 
This avoids costly operations like square root and division, making the design more efficient for hardware deployment.
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

// **Data Parameters**
#define OUT_C 3
#define H_OUT 10
#define W_OUT 12

typedef float fm_t;

// **BatchNorm Computation**
void batch_norm(
    fm_t outBuffer[OUT_C][H_OUT][W_OUT],
    fm_t inBuffer[OUT_C][H_OUT][W_OUT], 
    fm_t gamma[OUT_C], 
    fm_t beta[OUT_C],
    fm_t running_mean[OUT_C], 
    fm_t running_var[OUT_C], 
    fm_t epsilon = 1e-5
    ) {
    for (int oc = 0; oc < OUT_C; oc++) {
        fm_t mean = running_mean[oc];  
        fm_t var = running_var[oc];    
        fm_t inv_std = 1.0 / sqrt(var + epsilon);  // Compute inverse standard deviation (1 / sqrt(var + eps))

        for (int i = 0; i < H_OUT; i++) {
            for (int j = 0; j < W_OUT; j++) {
                fm_t norm_x = (inBuffer[oc][i][j] - mean) * inv_std;  // Normalize
                outBuffer[oc][i][j] = gamma[oc] * norm_x + beta[oc];  // Scale and shift
            }
        }
    }
}

// **Load Data from File**
void load_data(const std::string &filename, std::vector<fm_t> &data) {
    std::ifstream file(filename);
    if (!file) {
        std::cerr << "Unable to open file: " << filename << std::endl;
        exit(1);
    }
    fm_t value;
    while (file >> value) {
        data.push_back(value);
    }
}

int main() {
    // **Create Buffers**
    fm_t inBuffer[OUT_C][H_OUT][W_OUT];
    fm_t gamma[OUT_C], beta[OUT_C], running_mean[OUT_C], running_var[OUT_C];
    fm_t outBuffer[OUT_C][H_OUT][W_OUT];

    // **Load Data**
    std::vector<fm_t> in_data, gamma_data, beta_data, mean_data, var_data;
    load_data("data/inBuffer_bn2d.txt", in_data);
    load_data("data/gamma_bn2d.txt", gamma_data);
    load_data("data/beta_bn2d.txt", beta_data);
    load_data("data/running_mean_bn2d.txt", mean_data);
    load_data("data/running_var_bn2d.txt", var_data);

    // **Fill inBuffer**
    int index = 0;
    for (int oc = 0; oc < OUT_C; oc++)
        for (int i = 0; i < H_OUT; i++)
            for (int j = 0; j < W_OUT; j++)
                inBuffer[oc][i][j] = in_data[index++];

    // **Fill γ (scale), β (shift), running mean, and variance**
    for (int oc = 0; oc < OUT_C; oc++) {
        gamma[oc] = gamma_data[oc];
        beta[oc] = beta_data[oc];
        running_mean[oc] = mean_data[oc];
        running_var[oc] = var_data[oc];
    }

    // **Execute BatchNorm**
    batch_norm(inBuffer, outBuffer, gamma, beta, running_mean, running_var);

    // **Write Computation Results to File**
    std::ofstream file("data/out_cpp_bn2d.txt");
    if (!file) {
        std::cerr << "Unable to write file: data/out_cpp_bn2d.txt" << std::endl;
        return 1;
    }
    for (int oc = 0; oc < OUT_C; oc++)
        for (int i = 0; i < H_OUT; i++)
            for (int j = 0; j < W_OUT; j++)
                file << outBuffer[oc][i][j] << "\n";

    std::cout << "C++ computation complete, results saved to `data/out_cpp_bn2d.txt`" << std::endl;
    return 0;
}
